subsys/net/conn_mgr/conn_mgr_monitor.c

```diff
@@ -91,6 +143,537 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+		len = MIN(maxlen - 1, ptr - host);
+	} else {
+		len = strlen(host);
+		len = MIN(maxlen - 1, len);
+	}
+
+	memcpy(hostname, host, len);
+	hostname[len] = '\0';
+
+	return 0;
+}
+
+static void set_host_and_port(struct online_check_data *data)
+{
+	data->port = strstr(data->hostname_port, ":");
+	if (data->port != NULL) {
+		*data->port++ = '\0';
+	} else {
+		if (data->is_tls) {
+			data->port = "443";
+		} else {
+			data->port = "80";
+		}
+	}
+
+	data->host = data->hostname_port;
+}
+
+static int resolve_url(const char *url)
+{
+	struct sockaddr addr;
+	bool is_tls;
+	bool status;
+	int ret;
+
+	ret = get_hostname(url, online_check->hostname_port,
+			   MAX_HOSTNAME_LEN, &is_tls);
+	if (ret < 0) {
+		NET_DBG("Cannot find hostname from %s", URL);
+		return ret;
+	}
+
+	online_check->is_tls = is_tls;
+
+	/* Try to parse the address first */
+	status = net_ipaddr_parse(online_check->hostname_port,
+				  strlen(online_check->hostname_port),
+				  &addr);
+	if (status) {
+		/* We could parse the IP address */
+		memcpy(&online_check->hostaddr, &addr,
+		       sizeof(online_check->hostaddr));
+		online_check->hostaddr_valid = true;
+
+		set_host_and_port(online_check);
+	} else {
+		/* We need to do a DNS query */
+
+		set_host_and_port(online_check);
+
+		ret = resolve_hostname(online_check->host, online_check->port,
+				       SOCK_STREAM);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	if (online_check->hostaddr_valid) {
+		if (net_sin(&online_check->hostaddr)->sin_port == 0) {
+			if (online_check->is_tls) {
+				net_sin(&online_check->hostaddr)->sin_port = htons(443);
+			} else {
+				net_sin(&online_check->hostaddr)->sin_port = htons(80);
+			}
+		}
+	}
+
+	return 0;
+}
+
+static void response_cb(struct http_response *rsp,
+			enum http_final_call final_data,
+			void *user_data)
+{
+	uint16_t *status = user_data;
+
+	if (final_data == HTTP_DATA_MORE) {
+		NET_DBG("Partial data received (%zd bytes)", rsp->data_len);
+	} else if (final_data == HTTP_DATA_FINAL) {
+		NET_DBG("All the data received (%zd bytes)", rsp->data_len);
+	}
+
+	NET_INFO("Response status %s", rsp->http_status);
+
+	*status = rsp->http_status_code;
+}
+
+static int exec_http_query(struct net_if *iface, int sock)
+{
+	int32_t timeout = HTTP_WAIT_TIME;
+	struct http_request req;
+	uint16_t status;
+	int ret;
+
+	memset(&req, 0, sizeof(req));
+
+	req.method = HTTP_GET;
+	req.url = URL;
+	req.host = online_check->host;
+	req.port = online_check->port;
+	req.protocol = "HTTP/1.1";
+	req.response = response_cb;
+	req.recv_buf = online_check->http_recv_buf;
+	req.recv_buf_len = sizeof(online_check->http_recv_buf);
+
+	/* Make sure that we cannot access recv_buf data multiple times */
+	k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+
+	ret = http_client_req(sock, &req, timeout, &status);
+	if (ret < 0) {
+		NET_DBG("HTTP request failed (%d)", ret);
+		goto out;
+	}
+
+	/* Lets assume online also for 301 "Moved Permanently" */
+	if (status == 200 || status == 301) {
```

Comment:
Yes, it might be enough. I was thinking that perhaps the user of this API wants to actually connect so that the device gets registered somehow to some management system etc. This would be the "ET calls home" feature. To make that really work, the URL would need to be unique which is not atm.

Reply:
> To make that really work, the URL would need to be unique which is not atm.

This was such a low hanging fruit that I implemented the support for this. Application can specify the used URL at runtime in the latest version.

Reply:
I'm also not sure the "ET calls home" feature should really be implemented at the `conn_mgr` level anyways -- that feels squarely like application-specific behavior. IE, the application-writer (or, perhaps, cloud-library writer) should write their own "ET calls home", rather than configure `conn_mgr` to do one.

In fact, I do wonder what the intended use-case of the HTTP method in general is -- in theory, an application can get the same information by simply assuming connectivity is already there, attempting to connect, and accounting for the possibility of a connection failure/timeout, right?

So, to me, I hesitate to include such a sophisticated feature in `conn_mgr` -- it feels simpler, and more intuitive, to me, to let applications handle the rest themselves,

I am 100% on board with the ping, method, though, I think that could be useful

Reply:
The HTTP support is there because you might not have PING supported in all the networks. The "ET call home" feature is just a side effect that could be utilized if needed, it was not the main motivation for this support.
The point of this PR is that each and every application does not need to implement their own way to try to connect and figure out if they are connected to Internet.

Reply:
> The point of this PR is that each and every application does not need to implement their own way to try to connect and figure out if they are connected to Internet, or to what ever network they want to verify the connection.

I guess where I don't fully follow is that if the application needs to know this, its because the application wants to connect at some point

So why would an application use the HTTP connection check mode when they will get identical information from just connecting as usual whenever `conn_mgr` reports L4 up?

Like, why pre-check the connection, when the application could instead just simply connect?